### ISSUE TITLE
Java Integration

### ISSUE THREAD

**Issue by kienhaw** [fabriceci/RichFilemanager-JAVA#5]
Dear Sir,

I'm lost while integrating the Filemanager using Java connector. I've exactly followed all the instruction but I'm not so sure about this (refer to first attached image). Why is it returning "admin/rfm/home", where is the (JS) plugin's folder and how can I test if it is working with this java connector. I've been trying to launch the filemanager from CKEditor and it shows this (refer to second image attached) and I'm sure I am pointing to the right path (refer to third & fourth image attached). Last but not least, i would also like to raise a question about the API mentioned in your documentation, I can't quite get what is the API for. Maybe Sir you can provide me a demo of setting up the filemanager with Java connector. 

Sorry for any inappropriate sentences/request.

![image](https://user-images.githubusercontent.com/32449757/40092526-4b9f0258-58f0-11e8-8c10-474f1083ce05.png)

![image](https://user-images.githubusercontent.com/32449757/40092451-f1e44ec6-58ef-11e8-8fd6-7f4746b4e9b3.png)

![image](https://user-images.githubusercontent.com/32449757/40092647-e91a41fa-58f0-11e8-9a2e-3edc48326012.png)

![image](https://user-images.githubusercontent.com/32449757/40092476-0c1665cc-58f0-11e8-8538-2fe3fe754d24.png)

**Comment by fabriceci** [comment]
Hi,

I see many red color in your screenshot. You seem to have a problem with the Spring MVC framework more than the connector installation, you should before, read some docs about it. Do you really intent to use SPRING MVC in your project? It's not mandatory to use this framework with the connector.

I will try to help but you need to understand at least the basic of the framework.

FYI, the only line needed for the connector is the line 41 in your screenshot. You can add this line in all Java frameworks.

1) Don't add the public class AdminiFileManager... , just add the method in your current controller (MainController).

2) Why is it returning "admin/rfm/home"

Spring MVC controller return a string who lead to a view, so « admin/rfm/home » is the path of **_my_** view that displays the content of the [index.html](https://github.com/servocoder/RichFilemanager/blob/master/index.html) (RichFileManager file).

If you use Angular to route your app, serve the [index.html](https://github.com/servocoder/RichFilemanager/blob/master/index.html) directly in angular and delete this method. You only need the api method.

3) CkEditor integration

Before that, you need to be able to use the plugin in your browser by an URL. In my project the URL is "/admin/fileManager", but you can serve the content of the [index.html](https://github.com/servocoder/RichFilemanager/blob/master/index.html)l in the URL of your choice (angular router or, in server side in Spring MVC, or something else)

**Comment by kienhaw** [comment]
Thanks for the quick response Mr. Fabriceci! 

I believe there is no red color or errors in my project perhaps the only red color you saw in my screenshot was either class or datatype, and yeap my intention is to use Spring MVC to create filemanager project. I've removed the public class AdminFileManager (refer to screenshot below) and retain the api method. Let me know if I've missed any steps because I can't test the filemanager from my due to the error 404 not found. However, there is an error occurs when i'm trying to use the modal to open the filemanager (2nd screenshot below, X-Frame-Options error). But when I try to use window pop up it shows error 404 not found (last screenshot below)

![image](https://user-images.githubusercontent.com/32449757/40105786-9642a1b2-5926-11e8-87c7-c61f4f7e0e14.png)

![image](https://user-images.githubusercontent.com/32449757/40106137-8a5e1c22-5927-11e8-9f63-3f0dc5b7318f.png)

![image](https://user-images.githubusercontent.com/32449757/40106141-8ee1eaf8-5927-11e8-9776-f49ee6ded8da.png)

**Comment by kienhaw** [comment]
Dear Sir, 

1) However i managed to open up filemanager without using modal popup, but it shows this error. 

![image](https://user-images.githubusercontent.com/32449757/40594416-ad10698a-6261-11e8-8f5c-bc80fe6dbcc0.png)

Anyway, I did changed the {"/api"} to {"/admin/fileManager/api"} from my MainController but still no luck with the error below.

![image](https://user-images.githubusercontent.com/32449757/40594503-251880e8-6262-11e8-93d8-c1e2ceee9d65.png)

2) Regarding to the CkEditor integration, i did the routing of index.html using angular controller as shown in screenshot below.

![image](https://user-images.githubusercontent.com/32449757/40595070-15693d42-6265-11e8-8d5b-9c9604e4f31f.png)

And here is my latest file tree (where my filemanager and all controllers / ui contains).

![image](https://user-images.githubusercontent.com/32449757/40595127-64c1429a-6265-11e8-9c71-7f8318574212.png)

For filemanager.config.json 

![image](https://user-images.githubusercontent.com/32449757/40595587-1641f2e2-6268-11e8-9581-ded78cf4c101.png)

![image](https://user-images.githubusercontent.com/32449757/40595605-305d56d0-6268-11e8-87fb-59e6502bbda4.png)

**Comment by submediax** [comment]
Hi Kienhaw,

You might have defined the directory containing files in filemanager.config.properties to be `fileRoot=/admin/fileManager/api`.
If yes then check if this directory exists & user has permissions to access it.

**Comment by kienhaw** [comment]
Hi Submediax,

May i know the "filemanager.config.properties" you mentioned above is the one in the screenshot i've attached in the last comment? ("filemanager.config.default.properties"). If yes then i didn't define any directory (/admin/fileManager/api) in the properties files. 

/* Filemanager.config.default.properties */
![image](https://user-images.githubusercontent.com/32449757/40597734-cc332eec-6275-11e8-81f4-ba2c0da12eef.png)

/* Filemanager.config.json */
![image](https://user-images.githubusercontent.com/32449757/40597762-0d5f3438-6276-11e8-8d5d-38ca173488f3.png)


Moreover, the project now is developed locally (localhost) which shouldn't have any user restriction to any folder. 

Thanks for the quick response ! Cheers! Hope to hear from you soon.

**Comment by submediax** [comment]
Alright. _filemanager.config.default.properties_ should be definitely renamed into _filemanager.config.properties_ in order to configure Java connector. Afterwards define real path to your files dir by fileRoot property.

Good luck!

**Comment by kienhaw** [comment]
Thanks Oleg, by the way may i double confirm with you that fileRoot=/admin/fileManager/api (for example: myproject/admin (folder) /fileManager (folder) /api (folder) is a path and it isn't a URL right?

**Comment by kienhaw** [comment]
Hi Submediax,

I'm sorry for the late question, until now i'm still not sure what is the api and where is the api came from. 

![image](https://user-images.githubusercontent.com/32449757/40601087-44a6838c-6286-11e8-8ee3-d8f6db91e030.png)

In my api directory i don't have any file in it. 

![image](https://user-images.githubusercontent.com/32449757/40601257-ffbbd6e0-6286-11e8-8dec-9e4b4436c166.png)

I'm raising this because i keep getting the same error as shown in the screenshot below.

![image](https://user-images.githubusercontent.com/32449757/40601303-2ab33e60-6287-11e8-9481-04094a767ad8.png)

I have followed your instruction to rename filemanager.config.default.properties to filemanage.config.properties, and i've defined the directory containing the files in the properties to /admin/filemanager/api.

![image](https://user-images.githubusercontent.com/32449757/40601387-8fc8473c-6287-11e8-9c76-4c48b2ea2efb.png)

Unfortunately i'm still getting the same error.

**Comment by fabriceci** [comment]
You have a client app (angular) and a server API (spring MVC).

The "connectorURL" is the URL to reach your API (your controller), according your screenshot it's: /api

**Comment by submediax** [comment]
Hi again! By API we understand "Java connector for RFM plugin".
From one of your screenshots I saw `@RequestMapping(value = "/api")` so try to declare `"connectorUrl": "/api"`

**Comment by kienhaw** [comment]
Hi Fabriceci and submediax!

Thanks for the helps again ! I believe there's nothing to do with the properties/configuration, the controller itself is complaining because even if i try to directly access from the url (http://localhost:1234/api) it still giving the same error from the screenshot below. 

![image](https://user-images.githubusercontent.com/32449757/40601536-318ab960-6288-11e8-8870-0a74f70363f1.png)

![image](https://user-images.githubusercontent.com/32449757/40601669-a7fc86be-6288-11e8-994c-d971fda1de6c.png)

Here is my controller.

![image](https://user-images.githubusercontent.com/32449757/40601561-43f7ae0a-6288-11e8-9b07-f669e5fb375b.png)

connectorURL : 
![image](https://user-images.githubusercontent.com/32449757/40601572-54e0ac62-6288-11e8-9daf-300513cdefdc.png)

**Comment by fabriceci** [comment]
Try to debug the fm method (/api) with a break point, to see if you pass through. Maybe it's a server config issue.

**Comment by kienhaw** [comment]
Does this mean, it does pass through? Sorry, i'm not good on debugging. 
![image](https://user-images.githubusercontent.com/32449757/40602702-21bb1fee-628c-11e8-8652-b4e2958a9d90.png)

**Comment by kienhaw** [comment]
I'm sorry, refer to below screenshot for debugged info. 

![image](https://user-images.githubusercontent.com/32449757/40603700-4e59ee2e-628f-11e8-9dfc-a2a9f2bfcc08.png)

However, I finally managed to access filemanager without any error but the loading keeps spinning. In the console or log file isn't showing any errors, any idea sir?

![image](https://user-images.githubusercontent.com/32449757/40603769-857ba62c-628f-11e8-9b5f-007528413f73.png)

**Comment by kienhaw** [comment]
Dear Fabriceci and submediax,

Thank you all for helping me out here, I can finally open up the filemanager successfully. I've finally integrated the ckeditor with filemanager, but right now i'm facing a problem here. How can i upload the image to a certain directory/path ? as you can see in the screenshot below its showing /api?mode = xxx when i try to hit select button on the image in filemanager (unfortunately i can select but the window won't close)

![image](https://user-images.githubusercontent.com/32449757/40604343-4ffc5b48-6291-11e8-8bf9-e25771a989b0.png)

And right after i click on the select image, my application console showing this: 
![image](https://user-images.githubusercontent.com/32449757/40604422-8b8397e4-6291-11e8-8c6f-bb94c4f0139a.png)

Even if I've defined the directory in filemanager.config.properties but the image wasn't upload to the directory.

![image](https://user-images.githubusercontent.com/32449757/40604499-cb5a1244-6291-11e8-8343-84df6f17e873.png)

**Comment by kienhaw** [comment]
Regarding to the last comment that I've posted just now, how can i change the url of /api?mode=&path=? 

Right now, when i select the image from filemanager to ckeditor it forms a weird url which is right after /api. I wanted to change it to my own desired url, any suggestion where can i change the configuration? I'm able to track the the image that being uploaded now, but i just can't format the url. 

Ideal URL example:
http://domain/my/path/to/uploaded/image

P/S: I have tried to turn absolute path to true and defined fileRoot properties (uploads/), but when i select the image it shows ERROR on the image because the url is broken. EG: http://localhost:1234uploads/ instead of http://localhost:1234/uploads/

![image](https://user-images.githubusercontent.com/32449757/40609680-f823efd8-62a1-11e8-8e31-95440f403f8a.png)


Therefore the selected image cannot be displayed as the link wasn't available/right.

![image](https://user-images.githubusercontent.com/32449757/40608817-5b453f52-629f-11e8-99a4-42fda53b5069.png)

**Comment by fabriceci** [comment]
Yes, it’s the default behaviour, the client and the server should not necessarily be stored in the same server, and you have to pass by the API to get a file, which allows the keep the whole control.

But I think there is a parameter (and/or a hook, try _beforeCreatePreviewUrl_) in the [JavaScript plugin config](https://github.com/servocoder/RichFilemanager/wiki/Configuration-options) to change this (but I don't remember).

_When it works perfectly, it would be nice if you make suggestions to improve the documentation, so that others install it more easily._

**Comment by kienhaw** [comment]
Hi Fabriceci!

Thanks for the suggestion (Javascript plugin config), but i believe the client app i'm using now couldn't achieve the url format (eg: http://domain.com/folders/image). There will be a lot of work to do ( @RequestMapping method and retrieve image parameter) in order to form the url by the folder name and image name. Therefore, for what I have done now is that i added the protocol and hostname before the url (beforeSelectItem callbacks in filemanager.init.js) 

![image](https://user-images.githubusercontent.com/32449757/40695604-7ab3beee-63f4-11e8-9c3f-a37b95f60889.png)

And this was the outcome.

![image](https://user-images.githubusercontent.com/32449757/40695617-8704046a-63f4-11e8-8985-265eb43375c7.png)

Basically the documentation for installing java connector is simple and easy! The only issue that hangs me around is that I've imported HttpServletRequest in the controller page wrongly. Instead of ( import javax.servlet.http.HttpServletRequest; ), I imported (org.apache.catalina.servlet4preview.http.HttpServletRequest;) which causes me the internal server error.

And for the X-Frame-Options that i've encountered previously can be easily solve by adding http headers.disable() in the security configuration. (P/S: if developer are using Spring MVC)

![image](https://user-images.githubusercontent.com/32449757/40696564-ec2bb532-63f8-11e8-8c51-b7334fd855b7.png)



Thank you very much @fabriceci for your helps ! Much appreciated.
